### ISSUE TITLE
Required, Optional and Default Properties

### ISSUE THREAD

**Issue by ms14981** [ScottLogic/openapi-forge-java#49]
Currently, we don't differentiate between required and optional properties. See "Scenario: Required properties" in model.feature. We also are failing a couple of the cucumber tests because we are trying to check whether primitives are not null.

Java doesn't really have a concept of required or optional properties on a class or method. There are two ways we could represent it:

1) Use primitives to indicate required values, because they can't be null. Optional for us would therefore mean nullable. We'd have int vs Integer, double vs Double, char[] vs String, etc.
2) Use java.util.Optional to wrap optional values. Any values not wrapped by Optional are required, any that are become optional.
3) Don't distinguish between required vs optional. Box all values.

Consider a method with three parameters, one required, one optional and one optional with a default. This could be represented in the following ways:

1) Use primitives to indicate required values
```
public Response doSomething(char[] required, String optional, String optionalWithDefault) {
  if (optionalWithDefault == null) {
    // set to default value
  }
}
```

2) Use java.util.Optional to wrap optional values
```
public Response doSomething(String required, Optional<String> optional, Optional<String> optionalWithDefault) {
  if (optionalWithDefault.isEmpty()) {
    // set to default value
  }
}
```

3) Don't distinguish between required vs optional
```
public Response doSomething(String required, String optional, String optionalWithDefault) {
  if (optionalWithDefault == null) {
    // set to default value
  }
}
```

4) Don't distinguish between required vs optional and use overloaded methods to represent defaults.
```
public Response doSomething(String required, String optional, String optionalWithDefault) {
}
public Response doSomething(String required, String optional) {
  String optionalWithDefault = // set to default value
}
```

We need to decide which one makes most sense for both the generated code, and the cucumber code. The simplest solution to implement will be option 3 above (box everything).

**Comment by gkocak-scottlogic** [comment]
Here's my thought about this:

About option (1), if there is a case that the response method takes complex object and this complex object can be optional, I'm not sure if there is a way to distinguish the required / optional by using primitives.

```java
class InlineObject {
  String a,
  int b
}
...
public Response doSomething(InlineObject required, InlineObject optional) {
  // do stuff
}
```

You will still need to use null pointer logic for optional argument check.

---
(3) is solution that would definitely would fix the problem but boxing primitives doesn't sound appealing and it does not directly convey the optional information in the API in my opinion. But this might be the most _java_ way as well for the same reason.

```java
public Response doSomething(Integer required, Integer optional, Integer optionalWithDefault) {
  if (optionalWithDefault == null) {
    // set to default value
  }
  // optional integer present
  if (optional != null) {
    // do stuff
  }
}
```

---

(4) can lead to creating many overloading methods. Additionally, if we have multiple optionals of the same type like

```java
public Response doSomething(Integer required, Integer optional, Integer optional2) {}
```

I'm not sure if this logic can handle which optional parameter is present

---

Personally, I would recommend option (2) supplied with @NonNull notations for everything forbidding any null pointer logic. 

```java
public Response doSomething(@NonNull String required, @NonNull Optional<String> optional) {
  // stuff
}
```

I can see from some discussions that it's not very javaish [[1](https://stackoverflow.com/questions/31922866/why-should-java-8s-optional-not-be-used-in-arguments/39005452#39005452)]. There are some counter arguments about Optional being very plausible in the same thread as well (especially when we are parsing data into an object type).

There might be alternative solutions as well [[2](https://stackoverflow.com/questions/965690/how-do-i-use-optional-parameters-in-java)] such as using `Object...` or `Maps`. But I suspect they are a bit more old fashioned. But in this discussion, it's also pointed out that boxed primitives with null checkes are common practice optional parameters. So suggested option (2) seems like good way of tackling this in Java. 

---

To summarise, I think (2) or (3) are good options to consider for this situation.

**Comment by ms14981** [comment]
None of these solutions are perfect. I think we should go with (1), with one adaption. Char arrays can also be null, so for strings it makes sense to carry on using String for all properties; required, optional or default even though they can be null.

This would have the following advantages:

1. We are using primitives wherever possible, which is idiomatic in Java. [1]
2. We have compile time confidence that required properties cannot be null.
3. It's a straightforward distinction between properties in both generated code and tests.
  a. We'd need null checks for optional and default properties, but not required properties.
  b. We'd still need to box values that were used elsewhere, e.g. HttpResponse<int> is not valid.

@gkocak-scottlogic let's try to reach a shared decision next week!

[1] source required? There is some discussion [here](https://stackoverflow.com/questions/5199359/why-do-people-still-use-primitive-types-in-java) but it's by no means definitive

**Comment by ms14981** [comment]
Latest work on getting required values to be primitive and optional to be boxed: https://github.com/ScottLogic/openapi-forge-java/pull/59

**Comment by gkocak-scottlogic** [comment]
With the current PRs merged, currently we have generating primitives for the required values where possible. 

```java
public Response doSomething(int required, Integer optional) {
  if (optional == null){
    // stuff
  }
}
```

However, for object types, we don't distinguish in type references with required/optional while doing the necessary null checks for the optional parameters.

```java
public Response doSomething(String required, String optional) {
  if (optional == null){
    // stuff
  }
}
```

---

Next step is to add `@NonNull` annotation on required object parameters to incorporate this information which doesn't create any compile time or runtime checks. This will inform the user of the library about the our expectation of required properties being not null.


```java
public Response doSomething(@NonNull String required, String optional) {
  if (optional == null){
    // stuff
  }
}
```

**Comment by ms14981** [comment]
https://github.com/ScottLogic/openapi-forge-java/pull/71
